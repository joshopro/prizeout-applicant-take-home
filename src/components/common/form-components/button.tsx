import React from 'react';
import Classnames from 'classnames';
import PropTypes from 'prop-types';

import './button.less';

interface ButtonProps {
    ariaLabel: string;
    className?: string;
    color?: string;
    icon?: string;
    id?: string;
    isDisabled?: boolean;
    isInline?: boolean;
    isLink?: boolean;
    isLoading?: boolean;
    onClick?: () => void;
    size?: string;
    text?: string;
    type?: 'submit' | 'reset' | 'button' | undefined;
}

export const Button: React.FC<ButtonProps> = ({
    ariaLabel,
    className,
    color,
    isDisabled,
    icon,
    id,
    isInline,
    isLink,
    isLoading,
    onClick,
    size,
    text,
    type,
}): React.ReactElement => {
    const classes: string = Classnames(
        `button`,
        { [`button--${color}`]: color },
        { [`button--${size}`]: size },
        { 'button--disabled': isDisabled },
        { 'button--link': isLink },
        { 'button--inline': isInline },
        { 'button--loading': isLoading },
        className,
    );

    return (
        <button
            data-testid="button"
            onClick={onClick}
            id={id}
            type={type}
            className={classes}
            aria-label={ariaLabel}
            title={text}
            disabled={isDisabled}
        >
            {!isLoading && text && <span>{text}</span>}

            {icon && <i className={icon}></i>}

            {isLoading && (
                <div className="button__icon-wrapper">
                    <i data-testid="button-loading" className="fas fa-spinner fa-spin"></i>
                </div>
            )}
        </button>
    );
};

Button.propTypes = {
    ariaLabel: PropTypes.string.isRequired,
    className: PropTypes.string,
    color: PropTypes.string,
    icon: PropTypes.string,
    id: PropTypes.string,
    isDisabled: PropTypes.bool,
    isInline: PropTypes.bool,
    isLink: PropTypes.bool,
    isLoading: PropTypes.bool,
    onClick: PropTypes.func,
    size: PropTypes.string,
    text: PropTypes.string,
    type: PropTypes.oneOf(['submit', 'reset', 'button', undefined]),
};
