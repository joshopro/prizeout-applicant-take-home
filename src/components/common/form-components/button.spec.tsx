import { render, screen } from '../../../testing/test-utils';
import React from 'react';
import '@testing-library/jest-dom';

import { Button } from './button';

describe('Test Button Component', () => {
    const ariaLabel = `Aria label`;
    const color = `primary`;
    const isDisabled = true;
    const isInline = false;
    const isLink = false;
    const isLoading = false;
    const onClick = jest.fn();
    const size = `medium`;
    const text = `Button text`;
    const type = `submit`;

    // Matches snapshot
    test('Component matches snapshot', () => {
        const { asFragment } = render(
            <Button
                ariaLabel={ariaLabel}
                color={color}
                isDisabled={isDisabled}
                isInline={isInline}
                isLink={isLink}
                isLoading={isLoading}
                onClick={onClick}
                size={size}
                text={text}
                type={type}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    // Button text renders if not loading
    test('Button text renders on button if not loading', () => {
        render(
            <Button
                ariaLabel={ariaLabel}
                color={color}
                isDisabled={isDisabled}
                isInline={isInline}
                isLink={isLink}
                isLoading={isLoading}
                onClick={onClick}
                size={size}
                text={text}
                type={type}
            />,
        );

        expect(screen.getByText(text)).toBeVisible();
    });

    // Display button loading spinner when loading
    test('Button spinner is displayed if loading', () => {
        render(
            <Button
                ariaLabel={ariaLabel}
                color={color}
                isDisabled={isDisabled}
                isInline={isInline}
                isLink={isLink}
                isLoading={!isLoading}
                onClick={onClick}
                size={size}
                text={text}
                type={type}
            />,
        );

        expect(screen.getByTestId('button-loading')).toBeVisible();
    });

    // Append disabled class if disabled
    test('Disabled class if disabled state is true', () => {
        render(
            <Button
                ariaLabel={ariaLabel}
                color={color}
                isDisabled={isDisabled}
                isInline={isInline}
                isLink={isLink}
                isLoading={!isLoading}
                onClick={onClick}
                size={size}
                text={text}
                type={type}
            />,
        );

        expect(screen.getByTestId('button')).toHaveClass('button--disabled');
    });
});
