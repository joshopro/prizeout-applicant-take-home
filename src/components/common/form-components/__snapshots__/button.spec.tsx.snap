// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button Component Component matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-label="Aria label"
    class="button button--primary button--medium button--disabled"
    data-testid="button"
    disabled=""
    title="Button text"
    type="submit"
  >
    <span>
      Button text
    </span>
  </button>
</DocumentFragment>
`;
